snapcraft/plugins/_ros/ros2.py

```diff
@@ -0,0 +1,181 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2017 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import contextlib
+import os
+import logging
+import shutil
+import subprocess
+
+from snapcraft import (
+    repo,
+    sources,
+)
+
+logger = logging.getLogger(__name__)
+
+_ROS2_URL_TEMPLATE = (
+    'https://raw.githubusercontent.com/ros2/ros2/{version}/ros2.repos'
+)
+
+
+_INSTALL_TOOLS_STEP = 'install-tools'
+_FETCH_ROS2_STEP = 'fetch-ros2'
+_BUILD_ROS2_STEP = 'build-ros2'
+
+
+class Bootstrapper:
+    """Bootstrap ROS2 by building the underlay from source."""
+
+    def __init__(self, *, version, bootstrap_path, ubuntu_sources, project):
+        """Initialize bootstrapper.
+
+        :param str version: The ROS2 version to boostrap (e.g. release-beta3)
+        :param str bootstrap_path: Working directory for the bootstrap process
+        :param str ubuntu_sources: Ubuntu repositories from which dependencies
+                                   will be fetched.
+        :param project: Instance of ProjectOptions for project-wide settings
+        :type project: snapcraft.ProjectOptions
+        """
+        self._version = version
+        self._ubuntu_sources = ubuntu_sources
+        self._project = project
+
+        self._bootstrap_path = bootstrap_path
+        self._tool_dir = os.path.join(self._bootstrap_path, 'tools')
+        self._state_dir = os.path.join(self._bootstrap_path, 'state')
+        self._underlay_dir = os.path.join(self._bootstrap_path, 'underlay')
+        self._install_dir = os.path.join(self._bootstrap_path, 'install')
+        self._build_dir = os.path.join(self._bootstrap_path, 'build')
+        self._source_dir = os.path.join(self._underlay_dir, 'src')
+
+    def get_build_packages(self):
+        """Return the packages required for building the underlay."""
+
+        packages = ['cmake']
+
+        # Dependencies for FastRTPS
+        packages.extend(['libasio-dev', 'libtinyxml2-dev'])
+
+        # Dependencies for the rest of ros2
+        packages.extend([
+            'libopencv-dev', 'libpoco-dev', 'libpocofoundation9v5',
+            'libpocofoundation9v5-dbg', 'python3-dev', 'python3-empy',
+            'python3-nose', 'python3-pip', 'python3-setuptools',
+            'python3-yaml', 'libtinyxml-dev', 'libeigen3-dev'])
```

Comment:
Save a few statements by writing it like this:
```
packages = [
    'cmake',
    # Dependencies for FastRTPS
    'libasio-dev', 'libtinyxml2-dev',
    # Dependencies for the rest of ros2
    ...
]

Reply:
Sure thing.

Reply:
I was going to propose this change as well, but it is somewhat more readable the original way and those statements don't really add up to the slowness of the build in comparison

Reply:
I have no strong opinions here, but I already made the change and think the readability wasn't harmed. Please let me know if you feel otherwise and I can revert that commit.